services/QuillLMS/app/queries/prompt_feedback_history.rb

```diff
@@ -0,0 +1,54 @@
+class PromptFeedbackHistory
+
+    def self.run(activity_id)
+        promptwise_postprocessing promptwise_sessions(activity_id)
+    end
+
+    def self.promptwise_sessions(activity_id)
+        activity_sessions = ActivitySession.where(activity_id: activity_id).includes(:feedback_sessions)
+        feedback_session_uids = activity_sessions.reduce([]) {|memo, a_s| memo.concat a_s.feedback_sessions.pluck(:uid)}
+
+        sql = <<~SQL 
+          SELECT prompt_id, feedback_session_uid, count(*) as session_count, bool_or(optimal) as at_least_one_optimal, MAX(attempt) as attempt_cardinal
+          FROM feedback_histories
+          WHERE feedback_session_uid IN (#{feedback_session_uids.map{|e| "'#{e}'"}.join(',')})
+          GROUP BY prompt_id, feedback_session_uid
+        SQL
```

Comment:
I know that this breaks the pattern established in the last set of reports we built, but I wonder if this wouldn't be better if we did more ActiveRecord composition with this query.  I think it could let us avoid some of the awkwardness in the `WHERE` clause here.  This could be:
```ruby
FeedbackHistory.select("prompt_id, feedback_session_uid, count(*) as session_count, bool_or(optimal) as at_least_one_optimal, MAX(attempt) as attempt_cardinal")
  .where(feedback_session_uid: feedback_session_uids)
  .group(:prompt_id, :feedback_session_uid)
```
I think this will get you the same data and be a bit more ActiveRecord composable.  The big thing here is not having to do the awkward `.map.join` in the WHERE clause.

Reply:
I like this idea, but when I tried to implement, it appears that AREL does not accept `AS` statements. Weird, I know. Going to keep this for now, because not using AS to indicate intent is going to make this code confusing.